### ISSUE TITLE
Optimizing how we read the wikipedia dump

### ISSUE THREAD

**Issue by davidebbo** [OneZoom/tree-build#2]
The wikipedia dump file  `latest-all.json.bz2` (from https://dumps.wikimedia.org/wikidatawiki/entities/) is enormous:
- ~80GB compressed as bz2
- ~1.5TB uncompressed
- 100M lines (which are wikipedia entries)

Reading it efficiently is a hard problem. Today, we speed that up by:
- Building an index on the first run, which takes many hours (I didn't time it exactly)
- After that, we rely on the index to read it in parallel using multi-processing

On my Macbook M1, it still takes 65m to process it once we have the index.

I don't know that we can do much to read the entire file faster. But let's step back...

Out of those 100M entries, we ignore the wide majority of them because they don't have the `numeric-id` we care about:

```python
if not(line.startswith(b'{"type":"item"') and quick_byte_match.search(line)):
    continue
```

Specifically, we are skipping all but 3.6M entries, so we only look at 3.6% of the overall entries.

This offers a very simple optimization opportunity: on the first run, we can filter out the lines we don't need, and save the result as a separate file, e.g. latest-all-filtered.json.

We can actually do additional filtering by observing the following:
- Each entry has 5 sub-objects: `labels`, `descriptions`, `aliases`, `claims`, and `sitelinks`
- We never look at `descriptions` and `aliases`, so we can completely remove them
- We only care about a subset of the `claims`, so can can filter out the rest. Specifically, we care about at `{ "P31", "P685", "P846", "P850", "P1391", "P5055", "P830", "P141", "P627", "P961" }`
- From what I can see, we only look at the `en` entry under `labels`, so we can filter the rest out
- Seems we're using all the `sitelinks`, so we'd leave those alone

We can probably do more aggressive things, but with more complexity and diminishing returns.

If we do all the above, we end up with a file that has size 9.3GB **uncompressed**, which is 0.06% of the original uncompressed file. It's also a lot smaller than the original **compressed** size of 80GB. At that point, it is small enough that there is no real reason to compress it and have to deal with the bz2 complexity and runtime cost.

To summarize, the proposed algorithm is:
- On first run, instead of building a bz2 offset file, build an uncompressed filtered json file
- Any time after, just use the uncompressed filtered json directly, bypassing the original bz2 altogether.

In my quick test, the processing takes about 2.5 minutes (after we create the filtered file), compared to the current 65m (after we create the bz2 index). **And that's without multiple processing.** If we add multi-processing, it would likely be way under a minute.

Obviously, we'll need to pay the cost of creating the smaller file each time we get a new dump, but we're already paying a higher cost than that today when we build the index (which is more expensive that the filtering phase).

Please let me know what you think about this approach.

**Comment by davidebbo** [comment]
> The problem here is that we can't guarantee whether a common-name item or the taxon item to which it refers come first in the JSON dump.

@hyanwong Here is a possible idea:
- When we do the filtering phase on the original dump, initially save the vernacular and taxon items to two different files.
- At the end, put them together into one file by simply concatenating them (I assume with taxa before vernacular?).
- Then during tree generation, we'll be guaranteed never to encounter a vernacular item before encountering the taxa that it maps to.

Could that save us from doing multiple passes at all?

Though I'm actually not sure, because at the time we process the dump, we don't yet have the popularity. So even with the taxa/vernacular ordering guarantee, we would not yet know how to set `only_if_more_popular` for the override logic.

WDYT?

**Comment by davidebbo** [comment]
This optimization is in.
